### PR TITLE
make sure we our env file writer, parser and shell agree on things

### LINKED ISSUE #16788: hostnamed: incorrect shell escaping/unescaping for hostname saved to file
**systemd version the issue has been seen with**
> 142 thru 146 (and latest HEAD)

**Used distribution**
 > Debian and Fedora

**Expected behaviour you didn't see**
 > Setting hostname with `\`, `"`,  `` ` ``, `$`  characters should work.

**Unexpected behaviour you saw**
 > Setting hostname with `\` characters is replaced with escaped characters: `\\` and the other characters are escaped with `\`. I guess this has not happened before since most people use shell escaping with `hostnamectl` tool and quotation marks on the input, whereas over dbus api no external escaping is performed.

**Steps to reproduce the problem**

There are some issues with the escaping/unescaping of strings with the backslash character:
```
OK:
# echo "obbardc-laptop"
obbardc-laptop
# hostnamectl set-hostname --pretty 'obbardc-laptop'
# cat /etc/machine-info 
PRETTY_HOSTNAME=obbardc-laptop
# hostnamectl | grep hostname
   Static hostname: obbardc-laptop


# echo "obbardc\-laptop"
obbardc\-laptop
# hostnamectl set-hostname --pretty 'obbardc\-laptop'
# cat /etc/machine-info 
PRETTY_HOSTNAME="obbardc\\-laptop"
# hostnamectl | grep hostname
   Static hostname: obbardc-laptop
   Pretty hostname: obbardc\\-laptop

# echo 'PRETTY_HOSTNAME="obbardc\-laptop"' > /etc/machine-info
# hostnamectl | grep hostname
   Static hostname: obbardc-laptop
   Pretty hostname: obbardc\-laptop


# echo "obbardc-lap\top"
obbardc-lap\top
# hostnamectl set-hostname --pretty 'obbardc-lap\top'
# cat /etc/machine-info 
PRETTY_HOSTNAME="obbardc-lap\\top"
# hostnamectl | grep hostname
   Static hostname: obbardc-laptop
   Pretty hostname: obbardc-lap\\top

# echo 'PRETTY_HOSTNAME="obbardc-lap\top"' > /etc/machine-info
# hostnamectl | grep hostname
   Static hostname: obbardc-laptop
   Pretty hostname: obbardc-lap\top


In this case when saving `\\` should not be escaped to `\\`:
# echo "obbardc-lap\\top"
obbardc-lap\top
# hostnamectl set-hostname --pretty 'obbardc-lap\\top'
# cat /etc/machine-info 
PRETTY_HOSTNAME="obbardc-lap\\top"
# hostnamectl | grep hostname
   Static hostname: obbardc-laptop
   Pretty hostname: obbardc-lap\\top
# echo 'PRETTY_HOSTNAME="obbardc-lap\\top"' > /etc/machine-info
# hostnamectl | grep hostname
   Static hostname: obbardc-laptop
   Pretty hostname: obbardc-lap\\top



# echo "obbardc-lap\\\\top"
obbardc-lap\\top
# hostnamectl set-hostname --pretty 'obbardc-lap\\\\top'
# cat /etc/machine-info
PRETTY_HOSTNAME="obbardc-lap\\\\top"
# hostnamectl | grep hostname
   Static hostname: obbardc-laptop
   Pretty hostname: obbardc-lap\\\\top

# echo 'PRETTY_HOSTNAME="obbardc-lap\\\\top"' > /etc/machine-info
# hostnamectl | grep hostname
   Static hostname: obbardc-laptop
   Pretty hostname: obbardc-lap\\\\top
```

the patch below seems to fix this, but it's dirty and I am not sure what implications it has elsewhere (the tests still pass). I'd love to fix this bug with some direction.
```
diff --git a/src/basic/escape.h b/src/basic/escape.h
index 0b00b116ed..90ff84d7f1 100644
--- a/src/basic/escape.h
+++ b/src/basic/escape.h
@@ -12,14 +12,14 @@
 
 /* What characters are special in the shell? */
 /* must be escaped outside and inside double-quotes */
-#define SHELL_NEED_ESCAPE "\"\\`$"
+#define SHELL_NEED_ESCAPE "\"`$"
 
 /* Those that can be escaped or double-quoted.
  *
  * Stricly speaking, ! does not need to be escaped, except in interactive
  * mode, but let's be extra nice to the user and quote ! in case this
  * output is ever used in interactive mode. */
-#define SHELL_NEED_QUOTES SHELL_NEED_ESCAPE GLOB_CHARS "'()<>|&;!"
+#define SHELL_NEED_QUOTES SHELL_NEED_ESCAPE GLOB_CHARS "\\'()<>|&;!"
 
 /* Note that we assume control characters would need to be escaped too in
  * addition to the "special" characters listed here, if they appear in the
```

### PR DESCRIPTION
**Pull request by poettering**
Fixes #16788

### CODE REVIEW

File: src/basic/env-file.c [hunk 1]
```diff
@@ -209,17 +209,21 @@ static int parse_env_file_internal(
...
+                        if (strchr(SHELL_NEED_ESCAPE, c)) {
+                                /* If this is a char that needs escaping, just unescape it. */
                                 if (!GREEDY_REALLOC(value, value_alloc, n_value+2))
                                         return -ENOMEM;
-                                value[n_value++] = '"';
-                        } else if (!strchr(NEWLINE, c)) {
+                                value[n_value++] = c;
+                        } else if (c != '\n') {
+                                /* If other char than what needs escaping, keep the "\" in place, like the
+                                 * real shell does. */
```
Comment (keszybz) [inline src/basic/env-file.c:219]: I see that python does the same thing. I guess that's reasonable to follow their lead on this.

**Review by keszybz — APPROVED** [review]

**Comment by keszybz** [comment]
```

379/809 test-fileio                             FAIL     0.12 s (killed by signal 6 SIGABRT)

--- command ---
SYSTEMD_LANGUAGE_FALLBACK_MAP='/tmp/autopkgtest.IJo3EF/build.4ml/systemd/src/locale/language-fallback-map' SYSTEMD_KBD_MODEL_MAP='/tmp/autopkgtest.IJo3EF/build.4ml/systemd/src/locale/kbd-model-map' PATH='/tmp/autopkgtest.IJo3EF/build.4ml/systemd/build-deb:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin' /tmp/autopkgtest.IJo3EF/build.4ml/systemd/build-deb/test-fileio
--- stderr ---
Failed to read $container of PID 1, ignoring: Permission denied
Found container virtualization none.
Got: <one=BAR>
Got: <two=bar>
Got: <three=333
xxxx>
Got: <four=44\"44>
Got: <five=55"55FIVEcinco>
Got: <six=seis sechs sis>
Got: <seven=sevenval#nocomment>
Got: <eight=eightval #nocomment>
Got: <export nine=nineval>
Got: <ten=>
Got: <eleven=value>
Got: <twelve=\value>
Got: <thirteen=\value>
one=[BAR]
two=[bar]
three=[333
xxxx]
four=[44\"44]
five=[55"55FIVEcinco]
six=[seis sechs sis]
seven=[sevenval#nocomment]
eight=[eightval #nocomment]
export nine=[nineval]
ten=[nineval]
eleven=[value]
twelve=[\value]
thirteen=[\value]
sh: 1: source: not found
Assertion 'sz == strlen(value)' failed at src/test/test-fileio.c:162, function test_one_shell_var(). Aborting.
-------

400/809 test-env-file                           FAIL     0.12 s (killed by signal 6 SIGABRT)

--- command ---
SYSTEMD_LANGUAGE_FALLBACK_MAP='/tmp/autopkgtest.IJo3EF/build.4ml/systemd/src/locale/language-fallback-map' SYSTEMD_KBD_MODEL_MAP='/tmp/autopkgtest.IJo3EF/build.4ml/systemd/src/locale/kbd-model-map' PATH='/tmp/autopkgtest.IJo3EF/build.4ml/systemd/build-deb:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin' /tmp/autopkgtest.IJo3EF/build.4ml/systemd/build-deb/test-env-file
--- stderr ---
sh: 1: source: not found
Assertion 'sz == strlen(v)' failed at src/test/test-env-file.c:165, function test_write_and_load_env_file(). Aborting.
-------
```

**Comment by poettering** [comment]
hmm, is "source" a bashism? how does that work on whatever weird shell debian/ubuntu is using? christ...

**Comment by poettering** [comment]
Hmm, apparently "." is posix and "source" is bash. 

Amazing how Debian/Ubuntu can create so much extra work with bash/shell incompat just so that they can waste memory and disk  by shipping two shells instead of just one... So annoying

**Comment by poettering** [comment]
Force pushed a new version. Only change: `source` → `.` for those `popen()` shell lines.

**Comment by poettering** [comment]
not sure how I am supposed to read the newest results... it appears that that debian weirdo shell appears to read those env var files different than bash...

**Comment by poettering** [comment]
hmm, maybe the "-n" switch is unknown to that weirdo shell's echo command...

**Comment by poettering** [comment]
So this appears to pass now. Given the positive review and the minimal changes afterwards I'll set the green label.

(minimal changes afterwards were just `source` → `.` and `echo` → `/bin/echo`, both to deal with shell incompatibilities)

**Pull request merged**